### PR TITLE
Exception in ECN detection when ACK loss happens

### PR DESCRIPTION
**Pull request by gloinul**
Adding an exception to the ECN detection to handle ACK loss if that occurs. 

Fixes Issue #1481

### CODE REVIEW

File: draft-ietf-quic-transport.md [hunk 1]
```diff
@@ -1738,6 +1738,14 @@ network, an endpoint verifies the following when an ACK_ECN frame is received:
 
 * The increase in ECT(0) and ECT(1) counters MUST be no greater than the number
   of packets newly acknowledged that were sent with the corresponding codepoint.
+  
+* If the increase in the ECT(0), ECT(1) and CE counters are greater than the 
+  number of newly ACKed packets, and received ACK's smallest acknowledged or
```
Comment (ianswett) [inline draft-ietf-quic-transport.md]: I would just say "ACK's smallest acknowledged" and drop "or gap indicated packet number"

File: draft-ietf-quic-transport.md [hunk 2]
```diff
@@ -1738,6 +1738,14 @@ network, an endpoint verifies the following when an ACK_ECN frame is received:
 
 * The increase in ECT(0) and ECT(1) counters MUST be no greater than the number
   of packets newly acknowledged that were sent with the corresponding codepoint.
+  
+* If the increase in the ECT(0), ECT(1) and CE counters are greater than the 
+  number of newly ACKed packets, and received ACK's smallest acknowledged or 
+  gap indicated packet number is larger than previously largest acknowledged
```
Comment (ianswett) [inline draft-ietf-quic-transport.md]: How about: "...larger than the previous largest acknowledged, then it's likely acknowledgements were missed, and the above comparison MUST NOT be performed."

Comment (gloinul) [reply draft-ietf-quic-transport.md]: Implemented

File: draft-ietf-quic-transport.md (hunk 1 above)
Comment (gloinul) [reply draft-ietf-quic-transport.md]: Implemented.

File: draft-ietf-quic-transport.md [hunk 3]
```diff
@@ -1739,6 +1739,13 @@ network, an endpoint verifies the following when an ACK_ECN frame is received:
 * The increase in ECT(0) and ECT(1) counters MUST be no greater than the number
   of packets newly acknowledged that were sent with the corresponding codepoint.
 
+* If the increase in the ECT(0), ECT(1) and CE counters are greater than the
+  number of newly ACKed packets, and received ACK's smallest acknowledged is
```
Comment (ianswett) [inline draft-ietf-quic-transport.md]: ACKed -> acknowledged
received -> the received

File: draft-ietf-quic-transport.md [hunk 4]
```diff
@@ -1739,12 +1739,14 @@ network, an endpoint verifies the following when an ACK_ECN frame is received:
...
-  acknowledgements were missed, and the above comparison MUST NOT be
-  performed. Instead a new comparision point is stored by the sender so that
-  only changes after this point will be used in the future comaparisions.
+* If the increase in the ECT(0), ECT(1) and CE counters are greater than
+  the number of newly acknowledged packets, and the received ACK's
+  smallest acknowledged is larger than the previous largest acknowledged,
+  then it's likely acknowledgements were missed, and the above comparison
+  MUST NOT be performed. Instead a new comparision point is stored by the
+  sender so that only changes after this point will be used in the future
+  comaparisions.
```
Comment (ianswett) [inline draft-ietf-quic-transport.md]: Mikkel's right, comaparisons -> comparisons

File: draft-ietf-quic-transport.md [hunk 5]
```diff
@@ -1739,6 +1739,14 @@ network, an endpoint verifies the following when an ACK_ECN frame is received:
 * The increase in ECT(0) and ECT(1) counters MUST be no greater than the number
   of packets newly acknowledged that were sent with the corresponding codepoint.
 
+* If the increase in the ECT(0), ECT(1) and CE counters are greater than
+  the number of newly acknowledged packets, and the received ACK's
+  smallest acknowledged is larger than the previous largest acknowledged,
```
Comment (martinthomson) [inline draft-ietf-quic-transport.md]: As I pointed out, this isn't ideal and will lead to ECN being disabled in some cases.  I don't know how much we are going to want to worry about that though.

**Pull request merged**